src/command_buffer.rs

```diff
@@ -0,0 +1,225 @@
+
+    unsafe fn grow(
+        min_size: usize,
+        cursor: usize,
+        align: usize,
+        storage: NonNull<u8>,
+    ) -> (NonNull<u8>, Layout) {
+        let layout = Layout::from_size_align(min_size.next_power_of_two().max(64), align).unwrap();
+        let new_storage = NonNull::new_unchecked(alloc(layout));
+        ptr::copy_nonoverlapping(storage.as_ptr(), new_storage.as_ptr(), cursor);
+        (new_storage, layout)
+    }
+
+    unsafe fn add_inner(&mut self, ptr: *mut u8, ty: TypeInfo) {
+        let offset = align(self.cursor, ty.layout().align());
+        let end = offset + ty.layout().size();
+
+        if end > self.layout.size() || ty.layout().align() > self.layout.align() {
+            let new_align = self.layout.align().max(ty.layout().align());
+            let (new_storage, new_layout) = Self::grow(end, self.cursor, new_align, self.storage);
+            if self.layout.size() != 0 {
+                dealloc(self.storage.as_ptr(), self.layout);
+            }
+            self.storage = new_storage;
+            self.layout = new_layout;
+        }
+
+        let addr = self.storage.as_ptr().add(offset);
+        ptr::copy_nonoverlapping(ptr, addr, ty.layout().size());
+        self.info.push(ComponentInfo {
+            ty_info: ty,
+            offset,
+        });
+        self.cursor = end;
+    }
+
+    /// Buffer 'reserved_entity' with 'bundle'
+    pub fn spawn_at(&mut self, ent: Entity, bundle: impl DynamicBundle) {
+        let len = bundle.type_info().len();
+        unsafe {
+            bundle.put(|ptr, ty| self.add_inner(ptr, ty));
+        }
+        let begin = self.entities.last().map_or(0, |x| x.end);
+        let end = begin + len;
+        self.entities.push(EntityIndex {
+            entity: ent,
+            begin,
+            end,
+        });
+    }
+
+    /// Spawn every `entity` recorded with their components
+    ///
+    /// # Example
+    /// ```
+    /// # use hecs::*;
+    /// let mut world = World::new();
+    /// let a = world.reserve_entity();
+    /// let mut recorder = CommandBuffer::new();
+    /// recorder.spawn_at(a, (false,0.0));
+    /// recorder.run_on(&mut world);
+    /// assert!(world.contains(a));
+    /// ```
+    pub fn run_on(&mut self, world: &mut World) {
+        let mut mark = self.entities.len() - 1;
+
+        for index in self.entities.iter() {
+            self.info[index.begin..index.end].sort_unstable_by_key(|z| z.ty_info);
+        }
+
+        for _ in 0..self.entities.len() {
+            let (ent, comps) = self.build(mark);
+            world.spawn_at(ent, comps);
+            if mark != 0 {
+                mark -= 1;
+            }
+        }
+        self.clear();
+    }
+
+    fn build(&mut self, mark: usize) -> (Entity, ReadyBuffer<'_>) {
+        // Amount of components entity has
+        let len = self.entities[mark].end - self.entities[mark].begin;
+        self.ids
+            .extend(self.info[0..len].iter().map(|x| x.ty_info.id()));
```

Comment:
Why do we always use the first `len` entries here, regardless of which entity is being spawned?

Reply:
That's a ... relic from the past and shouldn't be there I am surprised it didn't break anything.